pkg/ocgormv2/callbacks.go

```diff
@@ -0,0 +1,231 @@
+package ocgormv2
```

Comment:
Probably the most annoying review comment to receive post-merge, is it possible to have a single package with various methods for V1 and V2 gorm? With the approach of copying the code to swap out a few parts of it I'm concerned what'll happen over time, will the two diverge unintentionally for instance.

Reply:
The difficulties with reusing the same signatures are caused by changes like:
* Constructions like `db.Callback().Create().Before("gorm:create").Register("instrumentation:before_create", c.beforeCreate)` now return an error
* Callbacks now have a signature `func(db *gorm.DB, operation string)` instead of `func(scope *gorm.Scope, operation string)`
* The way how the current table is retrieved has changed from `scope.TableName()` to `db.Statement.Table`
* A similar change for the current SQL script `scope.SQL` -> `db.Statement.SQL.String()`
* The new Gorm v2 supports Go contexts, eliminating the need to store contexts in the scope field as done in Gorm v1

I would say the commit https://github.com/hashicorp/go-gin-gorm-opencensus/pull/11/commits/ffb6081e6a2a7dbab0d41f03b742356e3e5fa29f provides the best overview of the changes required, it is based on the file copied from v1 and contains only the changes to support v2.

Reply:
Thanks for putting the effort into evaluating this review comment! I checked out the branch you built and [compared against main](https://github.com/hashicorp/go-gin-gorm-opencensus/compare/c8291d61e529640413224828394ee75267f140c2...30e5d8156471ca7430a3c6627e423312ee5fe21f) before this changeset merged.

The surface area to keep in sync between the approach we've taken here vs the one I was enquiring about is effectively the same. I don't think it's worth the version bump. I guess this dependency is unlikely to change often and if it breaks it only breaks a portion of observability. I'm okay with leaving as is unless you feel strongly otherwise.

Reply:
Thank you for reviewing it. I am looking forward to using it in our codebase. 👍